### PR TITLE
feat(Path): extend path definition

### PR DESCRIPTION
**Pull request by ymski**
## Description

Currently, CARET supports analysis from the initiating node's publish time to the terminating node's subscription callback start time. However, when evaluating the performance of an application, the following processing times can be analysed.
- Processing time until the initiating node publishes
- Processing time for the callback at the terminating node

This pull request extends the definition of Path. It also provides an option to include the above processing times.

If we want to include the processing time of the beginning (or end) node, use the following script.
```
target_path = app.get_path('target_path')
target_path.include_first_callback=True
target_path.include_last_callabck=True
```

## Related links

- https://tier4.atlassian.net/browse/T4PB-25958

## Pre-review checklist for the PR author

- [x] I've confirmed the [contribution guidelines](https://github.com/tier4/caret/blob/main/.github/CONTRIBUTING.md).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] (Optional) The PR has been properly tested with CARET_report verification.
- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: src/caret_analyze/infra/lttng/records_provider_lttng.py
[the remark this answers is not in the document]
Comment (ymski) [reply src/caret_analyze/infra/lttng/records_provider_lttng.py:1]: fixed in [81bab2a](https://github.com/tier4/CARET_analyze/pull/263/commits/81bab2a633f7606bf9a97fc787dbb3d257e2b706)

File: src/caret_analyze/infra/lttng/records_source.py
[the remark this answers is not in the document]
Comment (ymski) [reply src/caret_analyze/infra/lttng/records_source.py:1]: > Is changing of callback_records method necessary?

It is not necessary. remove it in [668d80e](https://github.com/tier4/CARET_analyze/pull/263/commits/668d80e4a0c5b7dd4d4cdb44d806820a8d534d1d)

> I think path_end_records method is not necessary. Caller can use callback_records insted. I guess the main difference between two methods is how argument of merge_sequential. (callback_records: inner, path_end_records: left_use_latest) I think inner is better for this case.

fixed in [63bcfde](https://github.com/tier4/CARET_analyze/pull/263/commits/63bcfde979ddae649c2c3e6722dee6ddb2b69b7f), [b244ac5](https://github.com/tier4/CARET_analyze/pull/263/commits/b244ac5566b8d19e120971389788a9a990a80706), [21facb1](https://github.com/tier4/CARET_analyze/pull/263/commits/21facb1ee9f659668a0c3033d983b7aa9dea6d81)

> It is necessary to consider intra-publish in path_beginning_records.

fixed in [ffeff8d](https://github.com/tier4/CARET_analyze/pull/263/commits/ffeff8da512b2e7beaaff6f4db2cdafde23dda00)

File: src/caret_analyze/runtime/path.py
[the remark this answers is not in the document]
Comment (ymski) [reply src/caret_analyze/runtime/path.py:1]: > I think the naming enable_beginning and enable_end are not clear. My suggestions are include_first_callback and include_last_callback. Internal method names like to_path_beginning_records and to_path_end_records look good to me.
It is better to add @property to enable_beginning and enable_end.

fixed in [ad0aa1b](https://github.com/tier4/CARET_analyze/pull/263/commits/ad0aa1bb1ca15ea0d45b493cd3ecd65bbec337d8).


> Is Path.column_name used?

`Path.column_name` is used in `PathBase.to_dataframe()`. As column_name refers to a column in a cached record `PathBase.__records_cached` (which cannot be referenced from a subclass), the method was added to the subclass. It was not desirable to have exactly the same method, so changes were made to the base class. fixed in [bb36b78](https://github.com/tier4/CARET_analyze/pull/263/commits/bb36b780833ab5bb2c266d27691a9b9e78c1dda9).

>Is comment avoid conflict column name? necessary?
> I think variable main_records is not necessary.
> It is better to clear __records_cache in clear_cache().

fixed in [a8d6100](https://github.com/tier4/CARET_analyze/pull/263/commits/a8d61006b8e0f3e1a1268af0d1f662b0a0789b2d).

**Comment by takeshi-iwanari** [comment]
@ymski 

I tested this PR using the following two passes with [this trace data](https://github.com/tier4/CARET_report/releases/download/20221227/autoware_launch_trace_20221227-150253_universe_rosbag.zip). You can use [the attached notebook](https://github.com/tier4/CARET_analyze/files/10988519/path.zip) to test:

- target_path_0: path in sensing component
- target_path_1: path in planning component

There are some issues with target_path_0. Could you confirm them?

- message_flow + include_first_callback(True) for target_path_0
  - !!! graph becomes empty  :smiling_face_with_tear: 
- message_flow + include_last_callback(True) for target_path_0
  - the last callback is nicely added :+1: 
- message_flow + include_first_callback(True) for target_path_1
  - the first callback is nicely added :+1: 
- message_flow + include_last_callback(True) for target_path_1
  - the last callback is nicely added :+1: 

- ResponseTime.to_best_case_timeseries + include_first_callback(True) for target_path_0
  - !!! dataframe becomes empty  :smiling_face_with_tear: 
- ResponseTime.to_best_case_timeseries + include_last_callback(True) for target_path_0
  - the last callback is nicely added :+1: 
- ResponseTime.to_best_case_timeseries + include_first_callback(True) for target_path_1
  - the first callback is nicely added :+1: 
- ResponseTime.to_best_case_timeseries + include_last_callback(True) for target_path_1
  - the last callback is nicely added :+1:

**Comment by ymski** [comment]
@takeshi-iwanari 
Thank you for reviewing.

> There are some issues with target_path_0. Could you confirm them?

Thank you for confirming that it works. The problem with the `target_path_0`, graph not being drawn has been fixed, it was due to the fact that intra_process was not supported. Can you check again when you have time?

**Comment by takeshi-iwanari** [comment]
@ymski 
Thank you for fixing it! I'll test new code.

Meanwhile, could you merge main to this PR (ymski:feat/extend_path)?
The following two PRs are recently merged to main branch. It looks there is no conflict and we decided that we will test/fix the combination of this function (include_first_callback(True)) + stacked bar graph later. However, it should be confirmed that there is no degradation before merging this PR to main. When checking degradation, response_time(include_first_callback(False)) should be the same as before.

- https://github.com/tier4/CARET_analyze/pull/265
- https://github.com/tier4/CARET_analyze/pull/259

**Comment by ymski** [comment]
> Meanwhile, could you merge main to this PR (ymski:feat/extend_path)?

The main branch has been merged into this pull request.

>  I'll test it, but I would be glad if you could also test it on your side, although there is no specific script for testing...

OK. I'll try to test them. It appears that at least the first callback should be tested for those using intra-process communication and those using inter-process communication.

**Comment by takeshi-iwanari** [comment]
@ymski 
I confirmed the followings for each component path in Autoware:

- Response time(include_first/last_callback=False) is exactly the same as response time calculated in main branch
  - -> There is no degradation
- Response time(include_first/last_callback=True) is almost the same as response time(include_first/last_callback=False) + callback latency at first/last node
  - -> The latency at first/last callback is added as expected
  - (I checked average value only. That's why I said "almost" here)
- Response time(include_first/last_callback=True) is almost the same as response time(include_first/last_callback=False) of the path(original path + preceding node + succeeding node)
  - -> The latency at first/last callback is added as expected
  - (callback latency is larger than duration time from subscribe to publish. Also, the second response time includes communication latency. So, these two response times cannot be exact the same value)

As for functionality, it looks okay :+1: 
Give me some time to look over code.

**Review by takeshi-iwanari — APPROVED** [review]
LGTM

File: src/caret_analyze/infra/lttng/records_source.py
[the remark this answers is not in the document]
Comment (takeshi-iwanari) [reply src/caret_analyze/infra/lttng/records_source.py:1]: @isp-uetsuki 
If you are satisfied with current code, could you resolve this conversion?

[the remark this answers is not in the document]
Comment (isp-uetsuki) [reply src/caret_analyze/infra/lttng/records_source.py:1]: @takeshi-iwanari 
I resolve this conversation. I am still reviewing test code, which will end early next week.

[the remark this answers is not in the document]
Comment (takeshi-iwanari) [reply src/caret_analyze/infra/lttng/records_source.py:1]: @isp-uetsuki 
Thank you!
I unchecked the following two check boxes. Please check them when you finish your review. 

- The PR has been properly tested.
- The PR has been reviewed.

File: src/test/runtime/test_node_path.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+                [
+                    ColumnValue('callback_start'),
+                    ColumnValue('xxx'),
+                    ColumnValue('pub'),
+                ]
+            )
+
+        assert records.equals(expected)
+
+    def test_cache(self, mocker):
```
Comment (isp-uetsuki) [inline src/test/runtime/test_node_path.py:147]: @ymski It is better to test caching of to_path_beginning_records() and to_path_end_records().

**Review by isp-uetsuki — APPROVED** [review]
LGTM

**Pull request merged**